### ISSUE TITLE
No Meta key support for F/Cursor keys

### ISSUE THREAD

**Issue by mountainstorm** [tmux/tmux#1645]
tmux: 2.8
terminal; Alacritty
macOS

I've configured Alacritty to produce meta key based cursor keys: https://invisible-island.net/xterm/ctlseqs/ctlseqs.html#h2-PC-Style-Function-Keys

Essentially this amounts to adding the following to the Alacritty key bindings:

- { key: Right, mods: Alt, chars: "\x1b[1;3C" }
- { key: Right, mods: Command, chars: "\x1b[1;9C" }

You can test this inside Alacritty, and see that you get different esc sequences with each key.

But when you run tmux, both key combos yield the "\x1b[1;3C" combination.  A quick look in the tmux source suggests it has an incomplete version of the following modifier table:

                    Code     Modifiers
                  ---------+---------------------------
                     2     | Shift
                     3     | Alt
                     4     | Shift + Alt
                     5     | Control
                     6     | Shift + Control
                     7     | Alt + Control
                     8     | Shift + Alt + Control
                     9     | Meta
                     10    | Meta + Shift
                     11    | Meta + Alt
                     12    | Meta + Alt + Shift
                     13    | Meta + Ctrl
                     14    | Meta + Ctrl + Shift
                     15    | Meta + Ctrl + Alt
                     16    | Meta + Ctrl + Alt + Shift
                  ---------+---------------------------

Adjusting xterm-keys appears to make no difference

**Comment by nicm** [comment]
In tmux, meta is alt, there is no additional meta modifier. Sending meta
in xterm is obscure and few applications use it (emacs can perhaps), I'm
not sure of the value of supporting it.

**Comment by nicm** [comment]
It's complicated because I don't think there is a practical way for us to
tell if we get Escape+Key whether the terminal is configured for that to be
Meta or Alt. At the moment we assume it is Meta; but then for function keys
and the mouse we assume Alt is Meta (as you have found). If we separated
them then we would need some way to know how the terminal is configured, or
we would break all existing configs where the default keys are Meta
bindings not Alt, that is we would need some way to tell if the terminal
has metaSendsEscape on or not. In xterm, we can turn it on or off with an
escape sequence but that seems a receipe for problems since so many
terminals pretend to be xterm but probably do not support this.

So it is a lot of faff and potentially extra options for something that
needs to be configured at the terminal and hardly anyone will use. How many
Meta keys do you need - can you configure them explicitly using the
user-keys option? For example something like:

set -s user-keys[10] "\e[1;9C"
bind -n User10 lsk

**Comment by mountainstorm** [comment]
So defining user-keys works for keys in tmux, and I'm using that; my problem is it doesn't help when I want to map the escape code to something in a process I run under tmux.

Assuming I've not bound them in tmux; for all the other F keys it works as expected; I generate \e[23;9~ and a process running in tmux sees \e[23;9~ but if I send \e[1;9C tmux translates it into \e[1;3C (changes the modifier).  I suspect there must be some code in tmux which magically changes the modifier number to 3 if it doesn't recognise it.

I'm happy to map escape codes to actions in tmux and the processes I use, but I need tmux to pass them unaltered.

FWIW and for others wanting to solve this; I have a workaround with is to map (in my terminal) all keys with meta modifiers to custom escape sequences of the form \e[9<scancode>;<modifier>~ It works as tmux doesn't mess with them and passes them through.

It's not a great solution for normal keys (although arguably better than the weird shift+esc scheme xterm uses) but it's truly horrid to have to remap cursors/F1-4 (with modifiers) to another code just because tmux won't pass them on un-altered.

In summary; I'd be happy with not mangling the esc sequences not ending in ~ I don't need full Meta support

**Comment by nicm** [comment]
Bind the user keys to send-keys.

**Comment by nicm** [comment]
I doesn't magically change it to 3 it changes it to 3 because 9 is meta and 3 is alt and for tmux meta and alt are the same.

**Comment by mountainstorm** [comment]
nice solution; it means a big config file, but at least I can autogenerate it.  Thanks
